### PR TITLE
feat(Checkbox): add component

### LINKED ISSUE #634: I am missing a checkbox component.
This feature request is to have a checkbox component which is having nicer styling than basic HTML checkbox element. It can look like as a toggle button. 

You can check how we implement it in Boolean renderer.

### PR DESCRIPTION
**Pull request by layershifter**
Fixes #634.

This PR adds a base implementation of `Checkbox` component with basic`checkboxBehavior`. Also adds `stardust-checkmark` icon that is used for `Checkbox` component.

### `<Checkbox checked />`

image

### `<Checkbox checked toogle />`

image

### Playground

checkbox

### CODE REVIEW

File: packages/react/src/themes/base/components/Checkbox/checkboxVariables.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  disabledToggleColor: string
+  disabledToggleBackground: string
+  disabledToggleBorderColor: string
+
+  padding: string
+}
+
+export default (siteVars: any): CheckboxVariables => ({
+  checkboxBackground: 'transparent',
+  checkboxBorderColor: siteVars.colors.grey[300],
```
Comment (mnajdova) [inline packages/react/src/themes/base/components/Checkbox/checkboxVariables.ts:43]: Let's try to use the color scheme here, as this is a fresh new component ;)

Comment (mnajdova) [reply packages/react/src/themes/base/components/Checkbox/checkboxVariables.ts:43]: I meant in teams theme :)

File: packages/react/src/themes/teams/index.tsx [hunk 2]
```diff
@@ -36,6 +36,7 @@ const themeIcons: ThemeIcons = Object.keys(svgIconsAndStyles as {
 
 const icons: ThemeIcons = {
   ...themeIcons,
+  'stardust-checkmark': themeIcons['stardust-checkmark'],
```
Comment (mnajdova) [inline packages/react/src/themes/teams/index.tsx]: Can't we just use the `accept` icon from teams theme?

Comment (layershifter) [reply packages/react/src/themes/teams/index.tsx]: Overlooked it, definitely it's better to use `accept` 👍

File: docs/src/components/ExampleSnippet/renderElementToJSX.ts [hunk 3]
```diff
@@ -14,7 +14,7 @@ const renderElementToJSX = (element: React.ReactNode, triggerErrorOnRenderFn: bo
       (typeof element.type === 'function' ? 'NoDisplayName' : element.type),
     showDefaultProps: false,
     showFunctions: true,
-    filterProps: ['accessibility'],
+    filterProps: ['accessibility', 'onClick', 'onChange'],
```
Comment (mnajdova) [inline docs/src/components/ExampleSnippet/renderElementToJSX.ts:17]: 👍  maybe in the future we can make this filtering smarter, but definitely now we need them

File: packages/react/src/components/Checkbox/Checkbox.tsx [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    },
+  }
+
+  getInitialAutoControlledState(): CheckboxState {
+    return { checked: false, isFromKeyboard: false }
+  }
+
+  handleClick = (e: React.MouseEvent | React.KeyboardEvent) => {
+    const { disabled } = this.props
+    const checked = !this.state.checked
```
Comment (mnajdova) [inline packages/react/src/components/Checkbox/Checkbox.tsx:119]: nit, could be moved in the if statement

File: packages/react/src/themes/base/components/Checkbox/checkboxVariables.ts (hunk 1 above)
Comment (mnajdova) [reply packages/react/src/themes/base/components/Checkbox/checkboxVariables.ts:43]: As we decided not to add the color scheme now and make the component red-lined, let's be sure to track this work

File: packages/react/src/components/Checkbox/Checkbox.tsx [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+  isFromKeyboard,
+  UIComponentProps,
+} from '../../lib'
+import { ComponentEventHandler, WithAsProp, ShorthandValue, withSafeTypeForAs } from '../../types'
+import Icon from '../Icon/Icon'
+import Text from '../Text/Text'
+import { Accessibility } from '../../lib/accessibility/types'
+import { checkboxBehavior } from '../../lib/accessibility'
+
+export interface CheckboxProps extends UIComponentProps, ChildrenComponentProps {
```
Comment (mnajdova) [inline packages/react/src/components/Checkbox/Checkbox.tsx:21]: Let's add conformat test for the new component.

File: packages/react/src/lib/accessibility/Behaviors/Checkbox/checkboxBehavior.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ * Adds role='checkbox'. This allows screen readers to handle the component as a checkbox button.
+ * Adds attribute 'aria-checked=true' based on the property 'checked'.
+ * Adds attribute 'tabIndex=0' to 'root' component's part.
+ */
+const checkboxBehavior: Accessibility = props => ({
+  attributes: {
+    root: {
+      'aria-checked': props.checked,
+      role: 'checkbox',
+      tabIndex: 0,
```
Comment (mnajdova) [inline packages/react/src/lib/accessibility/Behaviors/Checkbox/checkboxBehavior.ts:20]: Just to be sure, we should not add any attribute if the checkbox is disabled?

File: packages/react/src/themes/teams/components/Checkbox/checkboxStyles.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import { ComponentSlotStylesInput, ICSSInJSStyle } from '../../../types'
+import { CheckboxProps, CheckboxState } from '../../../../components/Checkbox/Checkbox'
+import { CheckboxVariables } from './checkboxVariables'
+import getBorderFocusStyles from '../../getBorderFocusStyles'
+
+const checkboxStyles: ComponentSlotStylesInput<CheckboxProps & CheckboxState, CheckboxVariables> = {
+  root: ({ props: p, variables: v, theme: t }): ICSSInJSStyle => ({
+    ...getBorderFocusStyles({
+      siteVariables: t.siteVariables,
```
Comment (mnajdova) [inline packages/react/src/themes/teams/components/Checkbox/checkboxStyles.ts:9]: It seems that no focus styles are visible. The component looks exactly the same when focused:
image

File: packages/react/src/themes/base/components/Checkbox/checkboxVariables.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  checkboxBorderRadius: pxToRem(4),
+  checkboxBorderWidth: pxToRem(1),
+  checkboxColor: 'transparent',
+  checkboxMargin: `0 ${pxToRem(12)} 0 0`,
+  checkboxPadding: '0',
+
+  toggleBackground: 'transparent',
+  toggleBorderColor: siteVars.colors.grey[300],
+  toggleBorderStyle: `solid`,
+  toggleBorderRadius: pxToRem(8),
```
Comment (mnajdova) [inline packages/react/src/themes/base/components/Checkbox/checkboxVariables.ts]: nit: can we make this value bigger, it looks better that way:
before pxToRem(8)
image

after pxToRem(999)
image

File: docs/src/components/ExampleSnippet/renderElementToJSX.ts (hunk 3 above)
Comment (layershifter) [reply docs/src/components/ExampleSnippet/renderElementToJSX.ts:17]: Yep, I thinking about it and how we can avoid this filtering at all

File: packages/react/src/lib/accessibility/Behaviors/Checkbox/checkboxBehavior.ts (hunk 6 above)
Comment (layershifter) [reply packages/react/src/lib/accessibility/Behaviors/Checkbox/checkboxBehavior.ts:20]: https://www.w3.org/TR/wai-aria-practices/examples/checkbox/checkbox-1/checkbox-1.html

I don't see this in the spec, so decided to skip for the initial implementation

File: packages/react/src/components/Checkbox/Checkbox.tsx (hunk 4 above)
Comment (layershifter) [reply packages/react/src/components/Checkbox/Checkbox.tsx:119]: Will keep as it is for now :)

File: packages/react/src/components/Checkbox/Checkbox.tsx (hunk 5 above)
Comment (layershifter) [reply packages/react/src/components/Checkbox/Checkbox.tsx:21]: Added, have to add an additional handler for `onChange`. Decided to keep it flat to be more obvious

File: packages/react/src/themes/teams/components/Checkbox/checkboxStyles.ts (hunk 7 above)
Comment (layershifter) [reply packages/react/src/themes/teams/components/Checkbox/checkboxStyles.ts:9]: Fixed 👍

File: packages/react/src/components/Checkbox/Checkbox.tsx [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+  renderComponent({ ElementType, classes, unhandledProps, styles, accessibility }) {
+    const { label, icon, toggle } = this.props
+
+    return (
+      <ElementType
+        className={classes.root}
+        onClick={this.handleClick}
+        onChange={this.handleChange}
+        onFocus={this.handleFocus}
```
Comment (mnajdova) [inline packages/react/src/components/Checkbox/Checkbox.tsx:143]: Was this the thing missing for the focus styles? :)

Comment (layershifter) [reply packages/react/src/components/Checkbox/Checkbox.tsx:143]: Yep :) No IDE highlight 😭

**Review by mnajdova — APPROVED** [review]

**Pull request merged**